Blocks/Persistence/persistence.h

```diff
@@ -44,13 +44,54 @@ namespace persistence {
 
 namespace impl {
 
+class ThreeStageMutex final {
+ public:
+  struct TopLevelScopedLock final {
+    explicit TopLevelScopedLock(ThreeStageMutex& parent) : guard_(parent.stage1_) {}
+    std::lock_guard<std::mutex> guard_;
+  };
+
+  struct ThreeStagesScopedLock final {
+    explicit ThreeStagesScopedLock(ThreeStageMutex& parent) : parent_(parent) { parent_.stage1_.lock(); }
+    void AdvanceToStageTwo() {
+      assert(stage_ == 1);
+      parent_.stage2_.lock();
+      parent_.stage1_.unlock();
+      stage_ = 2;
+    }
+    void AdvanceToStageThree() {
+      assert(stage_ == 2);
+      parent_.stage3_.lock();
+      parent_.stage2_.unlock();
+      stage_ = 3;
+    }
+    ~ThreeStagesScopedLock() {
+      assert(stage_ == 3);
+      parent_.stage3_.unlock();
+    }
+    int stage_ = 1;
+    ThreeStageMutex& parent_;
+  };
+
+  struct InnerLevelScopedUniqueLock {
```

Comment:
`final` as well?
As I understand, this one and `TopLevelScopedLock` are per-stage scoped locks. Maybe name them `StageThreeScopedLock` and `StageOneScopedLock` respectively?